In [1]:
import torch
import gc
torch.cuda.empty_cache()
gc.collect()


0

In [2]:
from setup import setup_src_path
print(setup_src_path())
import data.processed as processed
import config.config as config
import utils.setup as setup
import utils.functions as fn
from importlib import reload



['/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/mlm', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '', '/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages', '/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/modules']


2024-06-05 01:47:41.196694: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-06-05 01:47:41.402823: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-06-05 01:47:42.258413: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [5]:
from datasets import load_from_disk,concatenate_datasets


adapter_name="mlm_union_all_S"
dataset = load_from_disk(f"{config.Config.DATASETS_SAVE_PATH}/datasets")



In [21]:
reload(config)

    
filtered_target = dataset['train'].filter(lambda example: example['genre'] == "slate").shuffle(seed=42)
#filtered_target = shuffled_filtered_target.train_test_split(test_size=0.1)


filtered_test_target = dataset['validation_matched'].filter(lambda example: example['genre'] == "slate")
train_target = filtered_target
test_target = filtered_test_target
train_target[2]
 

{'promptID': 139803,
 'pairID': '139803n',
 'premise': "The word that is used around Doug is 'operator.",
 'premise_binary_parse': '( ( ( The word ) ( that ( is ( used ( around Doug ) ) ) ) ) ( ( ( is ` ) operator ) . ) )',
 'premise_parse': '(ROOT (S (NP (NP (DT The) (NN word)) (SBAR (WHNP (WDT that)) (S (VP (VBZ is) (VP (VBN used) (PP (IN around) (NP (NNP Doug)))))))) (VP (VBZ is) (`` `) (NP (NN operator))) (. .)))',
 'hypothesis': "People only choose to say the word 'operator' when talking to Doug",
 'hypothesis_binary_parse': "( People ( only ( choose ( to ( ( say ( the ( word ( ` ( operator ' ) ) ) ) ) ( when ( talking ( to Doug ) ) ) ) ) ) ) )",
 'hypothesis_parse': "(ROOT (S (NP (NNS People)) (VP (ADVP (RB only)) (VBP choose) (S (VP (TO to) (VP (VB say) (NP (DT the) (NN word) (`` `) (NN operator) ('' ')) (SBAR (WHADVP (WRB when)) (S (VP (VBG talking) (PP (TO to) (NP (NNP Doug))))))))))))",
 'genre': 'slate',
 'label': 1}

In [22]:
from adapters import AutoAdapterModel,init
from tqdm import tqdm

from transformers import AutoConfig, AutoModelForMaskedLM

mdlcfg = AutoConfig.from_pretrained(
    config.Config.MODEL_NAME,
 
)
model = AutoModelForMaskedLM.from_pretrained(
    config.Config.MODEL_NAME,
)
init(model)
reload(fn)
fn.print_trainable_parameters(model)

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertFor

trainable params: 109514298 || all params: 109514298 || trainable%: 100.0


In [23]:
from adapters import ConfigUnion, LoRAConfig, PrefixTuningConfig, SeqBnConfig,SeqBnInvConfig,AdapterConfig,LoRAConfig

union_config = ConfigUnion(
    PrefixTuningConfig(prefix_length=10),
    LoRAConfig(r=8, alpha=16),

    SeqBnInvConfig(reduction_factor=2),
)
model.add_adapter(adapter_name, config=union_config)

model.train_adapter([adapter_name])
model.active_adapters = adapter_name
print(model.adapter_summary())
fn.print_trainable_parameters(model)

Name                     Architecture         #Param      %Param  Active   Train
--------------------------------------------------------------------------------
mlm_union_all_S          union            17,539,712      16.107       1       1
--------------------------------------------------------------------------------
Full model                               108,891,648     100.000               0
trainable params: 18131840 || all params: 127054010 || trainable%: 14.2709702747674


In [24]:
from transformers import AutoTokenizer

reload(processed)
tokenizer = AutoTokenizer.from_pretrained(config.Config.TOKENIZER_NAME)

tokenized_dataset= processed.tokenize_dataset(train_target,tokenizer)
tokenized_dataset

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids'],
    num_rows: 77306
})

In [25]:
chunk_size = 128
# Slicing produces a list of lists for each feature
tokenized_samples = tokenized_dataset[444:470]

results = fn.group_texts(tokenized_samples, chunk_size)
for chunk in results["labels"]:
    print(f"'>>> Chunk length: {len(chunk)}'")

'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'


In [26]:
# we group texts and chunk them
lm_datasets = tokenized_dataset.map(fn.group_texts, batched=True,fn_kwargs={'chunk_size': chunk_size})
lm_datasets

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
    num_rows: 23975
})

In [27]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
samples = [lm_datasets[i] for i in range(2)]
for sample in samples:
    _ = sample.pop("word_ids")
dd=data_collator(samples)
for chunk in dd["input_ids"]:
    print(f"\n'>>> {tokenizer.decode(chunk)}'")


'>>> [CLS] notieri in america goes to a big public high school, but everyone goes to a high school governed by [MASK] hierarchy of popularity and cliques. [SEP] everyone has big public schools [MASK]. [SEP] [CLS] i am [MASK] gay. [SEP] i am gay. [SEP] [CLS] the word that is used around [MASK] is'[MASK]. [SEP] [MASK] harmon choose to say the [MASK]'operator'when talking to doug [SEP] [CLS] however, [MASK] [MASK] - bound colorado students take a [MASK] college entrance exam, making the [MASK] an unreliable measure of school quality. [SEP] some compute take [MASK] college entrance exams. [SEP] [CLS] their issuing of [MASK] - term buy or [MASK] - term hold'

'>>> recommendations obviously intensifies pressure on companies to meet and beat earnings expectations at [MASK] costs. [SEP] most companies will meet their earning expectations. [SEP] [CLS] well, my wife nearly [MASK] brownish to see [MASK] castigate [MASK] in public, no less on the internet [MASK] [SEP] my wife nearly fainted [MASK

In [28]:

downsampled_dataset = lm_datasets.train_test_split(
    test_size=0.1, seed=42
)
downsampled_dataset

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 21577
    })
    test: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 2398
    })
})

In [29]:

reload(fn)
trainer = fn.train_mlm_model(model,adapter_name,data_collator,tokenizer, downsampled_dataset['train'],downsampled_dataset['test'])

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/accelerate/accelerator.py:446: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(


In [30]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/75 [00:00<?, ?it/s]

>>> Perplexity: 16.82


In [31]:
trainer.train()

  0%|          | 0/13500 [00:00<?, ?it/s]

{'loss': 2.372, 'grad_norm': 1.9281458854675293, 'learning_rate': 9.866153846153846e-05, 'epoch': 1.0}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 2.021080255508423, 'eval_runtime': 4.6185, 'eval_samples_per_second': 519.212, 'eval_steps_per_second': 16.239, 'epoch': 1.0}
{'loss': 2.0975, 'grad_norm': 1.806840181350708, 'learning_rate': 9.347692307692308e-05, 'epoch': 2.0}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.9344379901885986, 'eval_runtime': 4.6057, 'eval_samples_per_second': 520.657, 'eval_steps_per_second': 16.284, 'epoch': 2.0}
{'loss': 2.0494, 'grad_norm': 1.766139030456543, 'learning_rate': 8.82923076923077e-05, 'epoch': 3.0}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.890414834022522, 'eval_runtime': 4.5854, 'eval_samples_per_second': 522.966, 'eval_steps_per_second': 16.356, 'epoch': 3.0}
{'loss': 1.9977, 'grad_norm': 1.708723783493042, 'learning_rate': 8.310769230769231e-05, 'epoch': 3.99}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.8649585247039795, 'eval_runtime': 4.6525, 'eval_samples_per_second': 515.426, 'eval_steps_per_second': 16.121, 'epoch': 4.0}
{'loss': 1.9578, 'grad_norm': 1.7465577125549316, 'learning_rate': 7.792307692307694e-05, 'epoch': 4.99}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.8359423875808716, 'eval_runtime': 4.6495, 'eval_samples_per_second': 515.755, 'eval_steps_per_second': 16.131, 'epoch': 5.0}
{'loss': 1.9404, 'grad_norm': 1.787013292312622, 'learning_rate': 7.273846153846155e-05, 'epoch': 5.99}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.8068499565124512, 'eval_runtime': 4.6331, 'eval_samples_per_second': 517.581, 'eval_steps_per_second': 16.188, 'epoch': 6.0}
{'loss': 1.9141, 'grad_norm': 1.7441260814666748, 'learning_rate': 6.756153846153846e-05, 'epoch': 6.99}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7910763025283813, 'eval_runtime': 4.6636, 'eval_samples_per_second': 514.194, 'eval_steps_per_second': 16.082, 'epoch': 7.0}
{'loss': 1.9041, 'grad_norm': 1.6249626874923706, 'learning_rate': 6.237692307692307e-05, 'epoch': 7.99}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.8059656620025635, 'eval_runtime': 4.6517, 'eval_samples_per_second': 515.511, 'eval_steps_per_second': 16.123, 'epoch': 8.0}
{'loss': 1.883, 'grad_norm': 1.8277870416641235, 'learning_rate': 5.71923076923077e-05, 'epoch': 8.99}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.797169804573059, 'eval_runtime': 4.673, 'eval_samples_per_second': 513.165, 'eval_steps_per_second': 16.05, 'epoch': 9.0}
{'loss': 1.865, 'grad_norm': 1.63650381565094, 'learning_rate': 5.200769230769231e-05, 'epoch': 9.99}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7679332494735718, 'eval_runtime': 4.6532, 'eval_samples_per_second': 515.348, 'eval_steps_per_second': 16.118, 'epoch': 10.0}
{'loss': 1.8462, 'grad_norm': 1.8389012813568115, 'learning_rate': 4.6823076923076926e-05, 'epoch': 10.98}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7357343435287476, 'eval_runtime': 4.6542, 'eval_samples_per_second': 515.23, 'eval_steps_per_second': 16.114, 'epoch': 11.0}
{'loss': 1.8509, 'grad_norm': 1.7277165651321411, 'learning_rate': 4.164615384615385e-05, 'epoch': 11.98}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7436230182647705, 'eval_runtime': 4.6346, 'eval_samples_per_second': 517.411, 'eval_steps_per_second': 16.183, 'epoch': 12.0}
{'loss': 1.827, 'grad_norm': 1.9196062088012695, 'learning_rate': 3.646153846153846e-05, 'epoch': 12.98}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7176058292388916, 'eval_runtime': 4.6677, 'eval_samples_per_second': 513.744, 'eval_steps_per_second': 16.068, 'epoch': 13.0}
{'loss': 1.8232, 'grad_norm': 1.8192845582962036, 'learning_rate': 3.1276923076923075e-05, 'epoch': 13.98}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7474795579910278, 'eval_runtime': 4.6729, 'eval_samples_per_second': 513.169, 'eval_steps_per_second': 16.05, 'epoch': 14.0}
{'loss': 1.8124, 'grad_norm': 1.6865789890289307, 'learning_rate': 2.6092307692307695e-05, 'epoch': 14.98}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7347441911697388, 'eval_runtime': 4.6649, 'eval_samples_per_second': 514.05, 'eval_steps_per_second': 16.077, 'epoch': 15.0}
{'loss': 1.8083, 'grad_norm': 1.7819081544876099, 'learning_rate': 2.090769230769231e-05, 'epoch': 15.98}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7041438817977905, 'eval_runtime': 4.6758, 'eval_samples_per_second': 512.854, 'eval_steps_per_second': 16.04, 'epoch': 16.0}
{'loss': 1.797, 'grad_norm': 1.7338995933532715, 'learning_rate': 1.5723076923076922e-05, 'epoch': 16.97}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7313199043273926, 'eval_runtime': 4.6396, 'eval_samples_per_second': 516.854, 'eval_steps_per_second': 16.165, 'epoch': 17.0}
{'loss': 1.7996, 'grad_norm': 1.688770055770874, 'learning_rate': 1.0546153846153848e-05, 'epoch': 17.97}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7430341243743896, 'eval_runtime': 4.6503, 'eval_samples_per_second': 515.668, 'eval_steps_per_second': 16.128, 'epoch': 18.0}
{'loss': 1.787, 'grad_norm': 2.0088672637939453, 'learning_rate': 5.361538461538461e-06, 'epoch': 18.97}


  0%|          | 0/75 [00:00<?, ?it/s]

{'eval_loss': 1.7445213794708252, 'eval_runtime': 4.6355, 'eval_samples_per_second': 517.313, 'eval_steps_per_second': 16.18, 'epoch': 19.0}


Overwriting existing adapter 'mlm_union_all_S'.


{'train_runtime': 1752.3706, 'train_samples_per_second': 246.261, 'train_steps_per_second': 7.704, 'train_loss': 1.911978156255235, 'epoch': 19.0}


TrainOutput(global_step=12825, training_loss=1.911978156255235, metrics={'train_runtime': 1752.3706, 'train_samples_per_second': 246.261, 'train_steps_per_second': 7.704, 'train_loss': 1.911978156255235, 'epoch': 19.0})

In [32]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/75 [00:00<?, ?it/s]

>>> Perplexity: 5.68


In [33]:
import config.config as config

trainer.model.save_adapter(f"{config.Config.ADAPTER_SAVE_PATH}/{adapter_name}", adapter_name)


: 

#### This is a report of mlm for Slate domain using whole genre as target and then splitt after tokenization with seed 42. I used computer metrix for preplexity and early stopping 
The results before:
>>> Perplexity: 16.82

The results after:
>>> Perplexity: 5.68